src/libguac/client.c

```diff
@@ -701,6 +739,120 @@ int guac_client_owner_supports_required(guac_client* client) {
+ *
+ * @param user
+ *     The user to send the notification to, which will be the owner of the
+ *     connection.
+ *
+ * @param data
+ *     The data provided to the callback, which is the user that is joining the
+ *     connection.
+ *
+ * @return
+ *     Zero if the message is sent successfully to the owner, otherwise
+ *     non-zero, cast as a void*.
+ */
+static void* guac_client_owner_notify_join_callback(guac_user* user, void* data) {
+
+    const guac_user* joiner = (const guac_user *) data;
+    int retval = 0;
+
+    char* owner = "owner";
+    if (user->info.username != NULL)
+        owner = strdup(user->info.username);
+
+    char* joinName = "anonymous";
+    if (joiner->info.username != NULL)
+        joinName = strdup(joiner->info.username);
+
+    guac_user_log(user, GUAC_LOG_DEBUG, "Notifying owner %s of %s joining.", owner, joinName);
+    
+    /* Send required parameters to owner. */
+    if (user != NULL) {
+        const char* args[] = { (const char*)joinName, NULL };
+        retval = guac_protocol_send_msg(user->socket, GUAC_MSG_CLIENT_JOINED, args);
+    }
+
+    else
+        retval = -1;
+
+    free(owner);
+    free(joinName);
+
+    return (void*) ((intptr_t) retval);
+
+}
+
+int guac_client_owner_notify_join(guac_client* client, guac_user* joiner) {
+
+    /* Don't send msg instruction if client does not support it. */
+    if (!guac_client_owner_supports_msg(client))
+        return -1;
+
+    return (int) ((intptr_t) guac_client_for_owner(client, guac_client_owner_notify_join_callback, joiner));
+
+}
+
+/**
+ * A callback function that is invokved by guac_client_owner_notify_leave() to
+ * notify the owner of a connection that another user has left the connection,
+ * returning zero if the message is sent successfully, or non-zero
+ * if an error occurs.
+ *
+ * @param user
+ *     The user to send the notification to, which will be the owner of the
+ *     connection.
+ *
+ * @param data
+ *     The data provided to the callback, which is the user that is leaving the
+ *     connection.
+ *
+ * @return
+ *     Zero if the message is sent successfully to the owner, otherwise
+ *     non-zero, cast as a void*.
+ */
+static void* guac_client_owner_notify_leave_callback(guac_user* user, void* data) {
+
+    const guac_user* quitter = (const guac_user *) data;
+
+    char* ownerName = "owner";
+    if (user->info.username != NULL)
+        ownerName = strdup(user->info.username);
+
+    char* quitterName = "anonymous";
+    if (quitter->info.username != NULL)
+        quitterName = strdup(quitter->info.username);
+
+    guac_user_log(user, GUAC_LOG_DEBUG, "Notifying owner %s of %s leaving.", ownerName, quitterName);
+    
+    /* Send required parameters to owner. */
+    if (user != NULL) {
+        const char* args[] = { (const char*)quitterName, NULL };
+        return (void*) ((intptr_t) guac_protocol_send_msg(user->socket, GUAC_MSG_CLIENT_LEFT, args));
+    }
+    
+    free(ownerName);
```

Comment:
Same here - you don't want to free string constants.

Reply:
Fixed this one, as well...